iptables: allow skipping conntrack on pods with hostNetwork: true

bersoare opened the pull request:
pods serving connections in the host network namespace might exhibit lower than expected performance under high load due to CT. this PR allows an operator to specify l4 protocol ports for which they want to skip CT when a pod is deployed in the host ns, avoiding the performance cost of connection tracking.

```release-note
Allow bypassing conntrack in pods with hostNetwork: true
```

bersoare committed:
loadbalancer: add helper to unformat a l4 addr

L4AddrFromString takes in a string formatted L4Addr (as in 443/tcp)
and returns an equivalent L4Addr struct

Signed-off-by: Bernardo Soares <bersoare@cisco.com>

bersoare committed:
iptables: handle notrack rules for pods in hostNetwork

extended iptables.Manager so it can install notrack rules for
ports used by pods with the hostNetwork flag set to true.

since the reference can come from many pods in the host ns, and they
all share the host ns in the data path, rules are only changed when
at least one reference to the port exists (if we're adding)
or the last reference of a port goes away (if we're deleting)

Signed-off-by: Bernardo Soares <bersoare@cisco.com>

bersoare committed:
iptables: extend reconciler to handle noTrackHostPorts functionality

handling no-track-host-ports events in the reconciler

Signed-off-by: Bernardo Soares <bersoare@cisco.com>

bersoare committed:
pod watcher: handle no-track-host-ports annotation on pod events

watcher looks for this newly added annotation and calls
iptables module to execute the iptables changes

Signed-off-by: Bernardo Soares <bersoare@cisco.com>

bersoare committed:
docs/examples: add doc and example on the no-track-host-ports configuration

adding an example nginx pod with the annotation
updated tuning.rst with notes on when it should be used

Signed-off-by: Bernardo Soares <bersoare@cisco.com>

borkmann commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -281,13 +314,15 @@ type Manager struct {
+	addNoTrackPod          chan reconciliationRequest[noTrackPodInfo]
+	delNoTrackPod          chan reconciliationRequest[noTrackPodInfo]
+	setNoTrackHostPorts    chan reconciliationRequest[noTrackHostPortsPodInfo]
+	removeNoTrackHostPorts chan reconciliationRequest[podAndNameSpace]
```
just nit: should we name it consistently `addNoTrackHostPorts`, `delNoTrackHostPorts`?

Also for the commit descriptions (e.g. https://github.com/cilium/cilium/pull/41364/commits/e5090ce719e21ab969140f0f28a8c233c92547f9), please format them in similar style with line break between 70-80 chars. See Cilium (or Linux kernel) git log for examples.

bersoare replied:
makes sense, i'll rename them

borkmann commented on Documentation/operations/performance/tuning.rst:
```diff
@@ -365,6 +365,20 @@ To enable the iptables connection-tracking bypass:
+
+.. note::
+    Only UDP and TCP transport protocols are supported with the network.cilium.io/no-track-host-ports annotation at the time of writing.
+
```
Nice work! Some high-level questions.. from the code I saw that most of the below we do handle, but just to make sure:

- If a user changes the `network.cilium.io/no-track-host-ports: "999/tcp,8123/tcp"` to a different set of ports, will the agent code be able to handle this and readjust the iptables rules?
- Similarly, if a Pod had no `network.cilium.io/no-track-host-ports` annotation, and a user adds it later on, do we handle it properly?
- Similarly, if a Pod had `network.cilium.io/no-track-host-ports` annotation, and a user deletes it later on, do we handle it properly?
- If a user changes the `network.cilium.io/no-track-host-ports: "999/tcp"` to `network.cilium.io/no-track-host-ports: "999/udp"`?
- Does the reconciler handles cases where there are leftover/stale iptables rules containing "cilium no-track-host-ports return traffic" comment but no Pod is using these ports?

What happens if the Pod has `network.cilium.io/no-track-host-ports` but its not in hostNetwork? Do we then ignore the annotation? Probably worth spilling a log warning and this is then ignored?

Did you double check that netfilter CT entries for these Pods are not created?

bersoare replied:
except for the last bullet (leftover state), ive tested all the other scenarios and they all worked as expected. to be sure, i'll repeat the latest and greatest commit (after the rebases) today. 
on the last bullet, how can i put a node on that state? would it be enough to just punch the rules there myself and see if they get cleaned up?

i'll make sure we have logging for that case you pointed out. 

and lastly, yea - CT entries dont show up

bersoare replied:
https://github.com/cilium/cilium/pull/41364/commits/ed52ebc3de8fc8b88534b534c6ce35d03b6439a3#diff-868649ca7ea4fe2118c75c38b5a72489ecb872edb946ae94ae118d3f8538ab2b adds tests for the cases of adding/changing/removing ports.
on the reconciler, i couldn't find a code test for verifying that the rules get deleted but i managed to test it on a running cluster - the rules are named properly and the reconciler includes the newly added statements in its actions.

borkmann commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1973,3 +2042,158 @@ func allEgressMasqueradeCmds(allocRange string, snatDstExclusionCIDR string,
+	byProto := groupL4AddrsByProto(p)
+
+	for proto, ports := range byProto {
+		if err := m.hostNoTrackMultiPorts(ip4tables, "-A", proto, ports); err != nil {
```
Should this say: `if m.sharedCfg.EnableIPv4 {` given users could run Cilium also in IPv6-only mode.

bersoare replied:
good point - will surely do that

borkmann commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1973,3 +2042,158 @@ func allEgressMasqueradeCmds(allocRange string, snatDstExclusionCIDR string,
+	byProto := groupL4AddrsByProto(p)
+
+	for proto, ports := range byProto {
+		if err := m.hostNoTrackMultiPorts(ip4tables, "-D", proto, ports); err != nil {
```
ditto

bersoare committed:
iptables: make iptables executables a member of iptables Manager struct

improve testability by allowing iptables executable to be passed
down the code path instead of using a global variable.

Signed-off-by: Bernardo Soares <bersoare@cisco.com>

bersoare committed:
iptables: add test cases for no-track-host-ports

adding test coverage for no-track-host-ports

Signed-off-by: Bernardo Soares <bersoare@cisco.com>

ldelossa commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1973,3 +2050,168 @@ func allEgressMasqueradeCmds(allocRange string, snatDstExclusionCIDR string,
+}
+
+// replaceNoTrackHostPortRules replaces noTrackHostPort rules on a state change. the new ruleset is added, and the previous one is removed.
+func (m *Manager) replaceNoTrackHostPortRules(oldPorts, newPorts map[lb.L4Type][]uint16) error {
```
We call this `replace` however, we are not really performing a `iptables -R` replacement of the rules. 

Is there any issue with this? IIUC correctly we are doing an add then removal. 
Is there any advantage to doing a `replace`. I don't recall if iptables performs replaces atomically or not.

bersoare replied:
decided to not use the replace command because it wants the rule number as an argument. that means we'd need to remember the rule numbers assigned by the kernel when we added them in order to run a replace command. it adds more coupling with the kernel state, and in my view it adds some level of complexity

bersoare replied:
i could change the function name to be more accurate on what it actually does underneath. at the time i thought that replace was equivalent to add-then-cleanup - but maybe it isnt

ldelossa replied:
Id just confirm if theres any benefit to replace.  If theres not, we can just ignore this. 

If there is a major benefit, you could make a new chain with only one rule, and replace that one rule. This is more complex so i dont think it should be done unless replace really is useful.

ldelossa replied:
> i could change the function name to be more accurate on what it actually does underneath.

cant hurt, but this wasnt a particular concern, more just wondering if a replace "gets" you anything.

ldelossa reviewed (approved):
Overall looks good to me. 

Just one question about "add/remove" vs "replace" of notrack rules.
